modules/core/src/main/java/org/apache/ignite/internal/processors/platform/binary/PlatformBinaryProcessor.java

```diff
@@ -126,16 +128,28 @@ public PlatformBinaryProcessor(PlatformContext platformCtx) {
             case OP_GET_TYPE: {
                 int typeId = reader.readInt();
 
-                try {
-                    String typeName = platformContext().kernalContext().marshallerContext()
-                        .getClassName(MarshallerPlatformIds.DOTNET_ID, typeId);
+                ClassNotFoundException err = null;
 
-                    writer.writeString(typeName);
-                }
-                catch (ClassNotFoundException e) {
-                    throw new BinaryObjectException(e);
+                for (byte platformId : new byte[] {DOTNET_ID, JAVA_ID}) {
+                    try {
+                        String typeName = platformContext().kernalContext().marshallerContext()
+                            .getClassName(platformId, typeId);
+
+                        writer.writeString(typeName);
+
+                        err = null;
+
+                        break;
+                    }
+                    catch (ClassNotFoundException e) {
```

Comment:
Let's not use exceptions for control flow - may be introduce an overload for `MarshallerContext.getClassName` that does not throw exceptions?

Reply:
I see your point, but if we introduce a method that not throws then we should return a tuple from it - T2(className, errorMsg).
Moreover, the second value of the tuple required in rare cases when we can't resolve the class name from type id.

AFAIU `getClassName` executed several times on every interaction between DotNet and Java so the creation of an extra object on each invocation will hurt the performance.

So I propose to keep changes as is. What do you think?

Reply:
1. The cost of throwing an exception is **a lot** higher than a tuple allocation.
2. `getClassName` is called from .NET once per type, then cached

I'm not concerned about performance too much here because of (2). Exception-based control flow is just dirty.

Reply:
Fixed according to your proposal.